File: ws_cobot_pjt/frontend/src/monitor/workArea.ts
import type { Profile, WorkAreaPolicy } from "./api";

export type WorkArea = {
  height: number;
  bottomRange: [number, number];
  topRange: [number, number];
  bottomExcluded: number;
  topExcluded: number;
  usableHeight: number;
};

// v: 윗면에서 아래(+), V: 바닥에서 위(+). 경로 좌표를 변경하지 않는다.
export function topToBottom(height: number, v: number) {
  return height - v;
}

function area(
  height: number | undefined,
  low: number,
  high: number,
): WorkArea | null {
  if (
    typeof height !== "number" ||
    ![height, low, high].every(Number.isFinite) ||
    height <= 0 ||
    low < 0 ||
    low >= high ||
    high > height
  )
    return null;
  return {
    height,
    bottomRange: [low, high],
    topRange: [topToBottom(height, high), topToBottom(height, low)],
    bottomExcluded: low,
    topExcluded: height - high,
    usableHeight: high - low,
  };
}

export function profileWorkArea(profile?: Profile): WorkArea | null {
  const s = profile?.payload.surface;
  if (!Array.isArray(s?.valid_v_range_mm) || s.valid_v_range_mm.length !== 2)
    return null;
  return area(s.height_mm, s.valid_v_range_mm[0], s.valid_v_range_mm[1]);
}

export function policyWorkArea(
  profile?: Profile,
  policy?: WorkAreaPolicy,
): WorkArea | null {
  const height = profile?.payload.surface.height_mm;
  const p = policy?.payload;
  if (
    !p ||
    p.contract !== "hmi-work-area-policy/1" ||
    typeof height !== "number" ||
    ![p.top_exclusion_mm, p.bottom_exclusion_mm].every(Number.isFinite) ||
    p.top_exclusion_mm < 0 ||
    p.bottom_exclusion_mm < 0
  )
    return null;
  return area(
    height,
    p.bottom_exclusion_mm,
    topToBottom(height, p.top_exclusion_mm),
  );
}

export function sameWorkArea(a: WorkArea | null, b: WorkArea | null) {
  return (
    !!a &&
    !!b &&
    Math.abs(a.height - b.height) < 1e-6 &&
    a.bottomRange.every((v, i) => Math.abs(v - b.bottomRange[i]) < 1e-6)
  );
}

export function mm(value: number) {
  return String(+value.toFixed(2));
}
export function rangeText(range: [number, number]) {
  return range.map(mm).join(" ~ ");
}

File: ws_cobot_pjt/backend/app/work_area.py
"""HMI 작업 영역 합의본. 좌표 노드의 입력 스냅샷을 덮어쓰지 않는다."""
import json
import math
from copy import deepcopy
from pathlib import Path

from .storage import digest, encoded


def load_policy():
    value = json.loads((Path(__file__).resolve().parents[1] / 'config/work_area.json').read_text())
    if value.get('contract') != 'hmi-work-area-policy/1':
        raise ValueError('지원하지 않는 HMI 작업 영역 기준입니다.')
    for key in ('top_exclusion_mm', 'bottom_exclusion_mm'):
        n = value.get(key)
        if type(n) not in (int, float) or not math.isfinite(n) or n < 0:
            raise ValueError(f'{key}는 0 이상의 유한한 수여야 합니다.')
    return value


def register_policy(store):
    """동일 설정은 같은 ID, 변경된 설정은 새 관리 자산으로 보존한다."""
    policy = load_policy()
    sha = digest(encoded(policy))
    with store.db() as db:
        row = db.execute("SELECT id FROM assets WHERE kind='hmi_work_area_policy' AND sha256=?",
                         (sha,)).fetchone()
    aid = row['id'] if row else store.put_json(policy, 'hmi_work_area_policy', 'work-area-policy.json')['id']
    store.read_asset(aid, sha)
    return dict(id=aid, sha256=sha, payload=policy)


def mock_profile_with_policy(profile, policy):
    """모의 형상에만 적용. ROS 프로파일은 경로 노드 제공본을 그대로 사용한다."""
    result = deepcopy(profile)
    height = result['surface']['height_mm']
    low = policy['bottom_exclusion_mm']
    high = height - policy['top_exclusion_mm']
    if not 0 <= low < high <= height:
        raise ValueError('제외 구간이 양초 높이를 벗어납니다.')
    result['surface']['valid_v_range_mm'] = [low, high]
    result['note'] = '모의 형상에 합의 작업 영역을 적용했습니다. 실측·실기 도달 가능 판정이 아닙니다.'
    return result
